### PR TITLE
Refactor normalize by cells sz

### PR DESCRIPTION
**Pull request by Sichao25**
- Replace duplicate operations with new DKM functions.
- Update docstrings.
- Add a test for normalization.

### CODE REVIEW

File: dynamo/configuration.py [hunk 1]
```diff
@@ -62,16 +63,18 @@ def gen_layer_pearson_residual_key(layer: str):
...
 
     def select_layer_data(adata: AnnData, layer: str, copy=False) -> pd.DataFrame:
-        """select layer data based on layer key. The default layer is X layer in adata.
-        For layer-like data such as X stored in adata.X (but not in adata.layers) and protein data specified by dynamo convention,
-        this utility provides an unified interface for selecting layer data with shape n_obs x n_var."""
+        """select layer data based on layer key. The default layer is X layer in
+        adata. For layer-like data such as X stored in adata.X (but not in
+        adata.layers) and protein data specified by dynamo convention, this
+        utility provides an unified interface for selecting layer data with
+        shape n_obs x n_var."""
```
Comment (Xiaojieqiu) [inline dynamo/configuration.py:70]: this utility provides an unified interface for selecting layer data.  
        The default layer is X layer in
        adata with shape n_obs x n_var. For protein data it selects adata.obsm["protein"] as specified by dynamo convention (the number of proteins are generally less than detected genes `n_var`).  For other layer data, select data based on layer key with shape n_obs x n_var

File: dynamo/configuration.py [hunk 2]
```diff
@@ -137,6 +140,35 @@ def allowed_X_layer_names():
...
+        if not isinstance(total_layers, list):
+            total_layers = [total_layers]
+        if len(set(total_layers).difference(_adata.layers.keys())) == 0:
+            total = None
+            for t_key in total_layers:
+                total = _adata.layers[t_key] if total is None else total + _adata.layers[t_key]
+            _adata.layers["_total_"] = total
+            if extend_layers:
+                layers.extend(["_total_"])
+        return total_layers, layers
```
Comment (Xiaojieqiu) [inline dynamo/configuration.py:172]: this function is not clear to me. Can you also document each argument? 
Are layers are valid layers in the adata while total_layers are the layers that should be used to calculate total RNA for collapsing?

File: dynamo/configuration.py [hunk 3]
```diff
@@ -137,6 +140,35 @@ def allowed_X_layer_names():
...
+        X_total_layers: bool = False,
+        splicing_total_layers: bool = False
+    ):
+        """ Get a list of excluded layers based on the provided arguments."""
+        excluded_layers = []
+        if not X_total_layers:
+            excluded_layers.extend(["X"])
+        if not splicing_total_layers:
+            excluded_layers.extend(["spliced", "unspliced"])
+        return excluded_layers
```
Comment (Xiaojieqiu) [inline dynamo/configuration.py:153]: can you document the meaning of each argument? 
"""Get a list of excluded layers based on the provided arguments.""" (remove the space before Get)

File: dynamo/preprocessing/preprocessor_utils.py [hunk 4]
```diff
@@ -1210,34 +1210,44 @@ def calc_sz_factor(
     use_all_genes_cells: bool = True,
     genes_use_for_norm: Union[List[str], None] = None,
 ) -> anndata.AnnData:
-    """Calculate the size factor of the each cell using geometric mean of total UMI across cells for a AnnData object.
+    """Calculate the size factor of the each cell using geometric mean of total
```
Comment (Xiaojieqiu) [inline dynamo/preprocessing/preprocessor_utils.py:1213]: geometric mean or median

File: dynamo/preprocessing/preprocessor_utils.py [hunk 5]
```diff
@@ -1210,34 +1210,44 @@ def calc_sz_factor(
...
-        genes_use_for_norm: A list of gene names that will be used to calculate total RNA for each cell and then the
-            size factor for normalization. This is often very useful when you want to use only the host genes to
-            normalize the dataset in a virus infection experiment (i.e. CMV or SARS-CoV-2 infection). Defaults to None.
+        round_exprs: whether the gene expression should be rounded into
+            integers. Defaults to False.
+        method: the method used to calculate the expected total reads / UMI used
+            in size factor calculation. Only `mean-geometric-mean-total` /
+            `geometric` and `median` are supported. When `median` is used,
+            `locfunc` will be replaced with `np.nanmedian`. Defaults to
+            "median".
```
Comment (Xiaojieqiu) [inline dynamo/preprocessing/preprocessor_utils.py:1237]: explain the difference between `mean-geometric-mean-total`, `geometric` and `median`

File: dynamo/preprocessing/preprocessor_utils.py [hunk 6]
```diff
@@ -1263,24 +1273,20 @@ def calc_sz_factor(
...
     layers = DKM.get_available_layer_keys(_adata, layers)
     if "raw" in layers and _adata.raw is None:
         _adata.raw = _adata.copy()
 
-    excluded_layers = []
-    if not X_total_layers:
-        excluded_layers.extend(["X"])
-    if not splicing_total_layers:
-        excluded_layers.extend(["spliced", "unspliced"])
+    excluded_layers = DKM.get_excluded_layers(
```
Comment (Xiaojieqiu) [inline dynamo/preprocessing/preprocessor_utils.py:1286]: add some annotation to make it clear that, when splicing_total_layers is False, we normalize spliced and unspliced RNA separately using each layer's size factors. Similary, when  X_total_layers is False, we normalize X (normally it corresponds to the spliced RNA or total RNA for a conventional scRNA-seq or labeling scRNA-seq) using its own's size factor

File: dynamo/configuration.py (hunk 2 above)
Comment (Sichao25) [reply dynamo/configuration.py]: Added a docstring for this helper function. The size factor normalization function is able to calculate size factors from customized layers. Given a list of total_layers, this helper function will calculate a temporary `_total_` layer. Layers are the list of layers users give in the size_factor normalization function.

**Pull request merged**